# 1.6 Таблица "Командировки", запросы на выборку

---

## Инициализация БД

In [1]:
%load_ext sql
import sqlalchemy
engine = sqlalchemy.create_engine('mysql://user:pass@localhost:3306/stepik')
%sql mysql://user:pass@localhost:3306/stepik

'Connected: user@stepik'

## Заполнение таблицы из CSV файла

In [2]:
# Чтение файла в DataFrame
import pandas as pd
file = 'tables/trip.csv'
df = pd.read_csv(file)
df

trip_id           name             city  per_diem  date_first   date_last
0         1   Баранов П.Е.           Москва       700  2020-01-12  2020-01-17
1         2  Абрамова К.А.      Владивосток       450  2020-01-14  2020-01-27
2         3   Семенов И.В.           Москва       700  2020-01-23  2020-01-31
3         4   Ильиных Г.Р.      Владивосток       450  2020-01-12  2020-02-02
4         5   Колесов С.П.           Москва       700  2020-02-01  2020-02-06
5         6   Баранов П.Е.           Москва       700  2020-02-14  2020-02-22
6         7  Абрамова К.А.           Москва       700  2020-02-23  2020-03-01
7         8   Лебедев Т.К.           Москва       700  2020-03-03  2020-03-06
8         9   Колесов С.П.      Новосибирск       450  2020-02-27  2020-03-12
9        10   Семенов И.В.  Санкт-Петербург       700  2020-03-29  2020-04-05
10       11  Абрамова К.А.           Москва       700  2020-04-06  2020-04-14
11       12   Баранов П.Е.      Новосибирск       450  2020-04-18  2020-05-04
12       13   Лебедев Т.К.            Томск       450  2020-05-20  2020-05-31
13       14   Семенов И.В.  Санкт-Петербург       700  2020-06-01  2020-06-03
14       15  Абрамова К.А.  Санкт-Петербург       700  2020-05-28  2020-06-04
15       16  Федорова А.Ю.      Новосибирск       450  2020-05-25  2020-06-04
16       17   Колесов С.П.      Новосибирск       450  2020-06-03  2020-06-12
17       18  Федорова А.Ю.            Томск       450  2020-06-20  2020-06-26
18       19  Абрамова К.А.      Владивосток       450  2020-07-02  2020-07-13
19       20   Баранов П.Е.          Воронеж       450  2020-07-19  2020-07-25

Создание схемы таблицы:

In [5]:
%%sql
  DROP TABLE IF EXISTS trip;

CREATE TABLE IF NOT EXISTS trip (
       trip_id    INT PRIMARY KEY AUTO_INCREMENT,
       name       VARCHAR(30),
       city       VARCHAR(25),
       per_diem   DECIMAL(8, 2),
       date_first DATE,
       date_last  DATE
);

 * mysql://user:***@localhost:3306/stepik
0 rows affected.
0 rows affected.


[]

In [6]:
# Запись данных в таблицу из DataFrame
types = {
    'trip_id'    : sqlalchemy.Integer(),
    'per_diem'   : sqlalchemy.Numeric(precision=8, scale=2),
    'amount'     : sqlalchemy.Integer(),
    'date_first' : sqlalchemy.Date(),
    'date_last'  : sqlalchemy.Date()
}
df.to_sql('trip', con=engine, index=False, if_exists='append', dtype=types, method='multi')

***

# Упражнения

### Задание 1

Вывести из таблицы `trip` информацию о командировках тех сотрудников, фамилия которых заканчивается на букву «а»,  
в отсортированном по убыванию даты последнего дня командировки виде.  
В результат включить столбцы `name, city, per_diem, date_first, date_last`.

In [15]:
%%sql
SELECT name, city, per_diem, 
       date_first, date_last
  FROM trip
 WHERE name LIKE '%а _._.'
 ORDER BY date_last DESC

 * mysql://user:***@localhost:3306/stepik
7 rows affected.


[('Абрамова К.А.', 'Владивосток', Decimal('450.00'), datetime.date(2020, 7, 2), datetime.date(2020, 7, 13)),
 ('Федорова А.Ю.', 'Томск', Decimal('450.00'), datetime.date(2020, 6, 20), datetime.date(2020, 6, 26)),
 ('Абрамова К.А.', 'Санкт-Петербург', Decimal('700.00'), datetime.date(2020, 5, 28), datetime.date(2020, 6, 4)),
 ('Федорова А.Ю.', 'Новосибирск', Decimal('450.00'), datetime.date(2020, 5, 25), datetime.date(2020, 6, 4)),
 ('Абрамова К.А.', 'Москва', Decimal('700.00'), datetime.date(2020, 4, 6), datetime.date(2020, 4, 14)),
 ('Абрамова К.А.', 'Москва', Decimal('700.00'), datetime.date(2020, 2, 23), datetime.date(2020, 3, 1)),
 ('Абрамова К.А.', 'Владивосток', Decimal('450.00'), datetime.date(2020, 1, 14), datetime.date(2020, 1, 27))]

### Задание 2

Вывести в алфавитном порядке фамилии и инициалы тех сотрудников, которые были в командировке в Москве.

In [18]:
%%sql
SELECT DISTINCT name 
  FROM trip
 WHERE city = "Москва"
 ORDER BY name

 * mysql://user:***@localhost:3306/stepik
5 rows affected.


[('Абрамова К.А.',),
 ('Баранов П.Е.',),
 ('Колесов С.П.',),
 ('Лебедев Т.К.',),
 ('Семенов И.В.',)]

### Задание 3

Для каждого города посчитать, сколько раз сотрудники в нем были.  
Информацию вывести в отсортированном в алфавитном порядке по названию городов.  
Вычисляемый столбец назвать **Количество**.

In [32]:
%%sql
SELECT city, COUNT(*) AS Количество 
  FROM trip
 GROUP BY 1
 ORDER BY 1

 * mysql://user:***@localhost:3306/stepik
6 rows affected.


[('Владивосток', 3),
 ('Воронеж', 1),
 ('Москва', 7),
 ('Новосибирск', 4),
 ('Санкт-Петербург', 3),
 ('Томск', 2)]

## Оператор `LIMIT`

Для ограничения вывода записей в SQL используется оператор `LIMIT` , после которого указывается количество строк.  
Результирующая таблица будет иметь количество строк не более указанного после `LIMIT` .  
`LIMIT` размещается после раздела `ORDER BY`.

Как правило, этот оператор используется, чтобы отобрать заданное количество отсортированных строк результата запроса. 

**Пример**  
Вывести информацию о первой  командировке из таблицы `trip` .  
"Первой" считать командировку с самой ранней датой начала.

In [25]:
%%sql
SELECT *
  FROM trip
 ORDER BY date_first
 LIMIT 1

 * mysql://user:***@localhost:3306/stepik
1 rows affected.


[(1, 'Баранов П.Е.', 'Москва', Decimal('700.00'), datetime.date(2020, 1, 12), datetime.date(2020, 1, 17))]

**Важно.** Оператор `LIMIT` нужно использовать очень осторожно.  
Например, если бы в таблице `trip` было несколько командировок с одинаковой датой начала, этот запрос работал бы НЕВЕРНО.  
Это связано с тем, что заранее не известно точное значение таких командировок.

### Задание 4

Вывести два города, в которых чаще всего были в командировках сотрудники.  
Вычисляемый столбец назвать **Количество**.

In [33]:
%%sql
SELECT city, COUNT(*) AS Количество 
  FROM trip
 GROUP BY city
 ORDER BY Количество DESC
 LIMIT 2

 * mysql://user:***@localhost:3306/stepik
2 rows affected.


[('Москва', 7), ('Новосибирск', 4)]

### Задание 5, `DATEDIFF()`

Вывести информацию о командировках во все города  
кроме Москвы и Санкт-Петербурга (фамилии и инициалы сотрудников, город ,  длительность командировки в днях, при этом первый и последний день относится к периоду командировки).  
Последний столбец назвать `Длительность`. Информацию вывести в упорядоченном по убыванию длительности поездки, а потом по убыванию названий городов (в обратном алфавитном порядке).

#### Немного теории
Для вычитания двух дат используется функция `DATEDIFF(дата_1, дата_2)`, результатом которой является количество дней между **дата_1** и **дата_2**. Например,

>`DATEDIFF('2020-04-01', '2020-03-28')=4`  
`DATEDIFF('2020-05-09','2020-05-01')=8`  
`DATEDIFF(date_last, date_first)`

In [41]:
%%sql
SELECT name, city, 
       DATEDIFF(date_last, date_first) + 1 AS Длительность
  FROM trip
 WHERE city NOT IN ('Москва', 'Санкт-Петербург')
 ORDER BY Длительность DESC, 
          city DESC

 * mysql://user:***@localhost:3306/stepik
10 rows affected.


[('Ильиных Г.Р.', 'Владивосток', 22),
 ('Баранов П.Е.', 'Новосибирск', 17),
 ('Колесов С.П.', 'Новосибирск', 15),
 ('Абрамова К.А.', 'Владивосток', 14),
 ('Лебедев Т.К.', 'Томск', 12),
 ('Абрамова К.А.', 'Владивосток', 12),
 ('Федорова А.Ю.', 'Новосибирск', 11),
 ('Колесов С.П.', 'Новосибирск', 10),
 ('Федорова А.Ю.', 'Томск', 7),
 ('Баранов П.Е.', 'Воронеж', 7)]

### Задание 6

Вывести информацию о командировках сотрудника(ов), которые были самыми короткими по времени.  
В результат включить столбцы `name`, `city`, `date_first`, `date_last`.

In [51]:
%%sql
SELECT name, city, date_first, date_last
  FROM trip
 WHERE DATEDIFF(date_last, date_first) = 
       (SELECT MIN(DATEDIFF(date_last, date_first))
          FROM trip
       )

 * mysql://user:***@localhost:3306/stepik
1 rows affected.


[('Семенов И.В.', 'Санкт-Петербург', datetime.date(2020, 6, 1), datetime.date(2020, 6, 3))]

### Задание 7. `MONTH()`

Вывести информацию о командировках, начало и конец которых относятся к одному месяцу (год может быть любой).  
В результат включить столбцы `name`, `city`, `date_first`, `date_last` .  
Строки отсортировать сначала  в алфавитном порядке по названию города, а затем по фамилии сотрудника .

#### Немного теории
Для того, чтобы выделить номер месяца из даты используется функция `MONTH(дата)` .  
Например, `MONTH('2020-04-12') = 4` .  
Если определяется месяц для  значений столбца `date_first`, то используется запись `MONTH(date_first)`

In [54]:
%%sql
SELECT name, city, date_first, date_last
  FROM trip
 WHERE MONTH(date_first) = MONTH(date_last)
 ORDER BY city, name

 * mysql://user:***@localhost:3306/stepik
13 rows affected.


[('Абрамова К.А.', 'Владивосток', datetime.date(2020, 1, 14), datetime.date(2020, 1, 27)),
 ('Абрамова К.А.', 'Владивосток', datetime.date(2020, 7, 2), datetime.date(2020, 7, 13)),
 ('Баранов П.Е.', 'Воронеж', datetime.date(2020, 7, 19), datetime.date(2020, 7, 25)),
 ('Абрамова К.А.', 'Москва', datetime.date(2020, 4, 6), datetime.date(2020, 4, 14)),
 ('Баранов П.Е.', 'Москва', datetime.date(2020, 1, 12), datetime.date(2020, 1, 17)),
 ('Баранов П.Е.', 'Москва', datetime.date(2020, 2, 14), datetime.date(2020, 2, 22)),
 ('Колесов С.П.', 'Москва', datetime.date(2020, 2, 1), datetime.date(2020, 2, 6)),
 ('Лебедев Т.К.', 'Москва', datetime.date(2020, 3, 3), datetime.date(2020, 3, 6)),
 ('Семенов И.В.', 'Москва', datetime.date(2020, 1, 23), datetime.date(2020, 1, 31)),
 ('Колесов С.П.', 'Новосибирск', datetime.date(2020, 6, 3), datetime.date(2020, 6, 12)),
 ('Семенов И.В.', 'Санкт-Петербург', datetime.date(2020, 6, 1), datetime.date(2020, 6, 3)),
 ('Лебедев Т.К.', 'Томск', datetime.date(2020, 5, 20), datetime.date(2020, 5, 31)),
 ('Федорова А.Ю.', 'Томск', datetime.date(2020, 6, 20), datetime.date(2020, 6, 26))]

### Задание 8. `MONTHNAME()`

Вывести название месяца и количество командировок для каждого месяца.  
Считаем, что командировка относится к некоторому месяцу, если она началась в этом месяце.  
Информацию вывести сначала в отсортированном по убыванию количества, а потом в алфавитном порядке по названию месяца виде.  
Название столбцов – **Месяц** и **Количество**.

#### Немного теории
1. Для того, чтобы выделить название месяца из даты используется функция `MONTHNAME(дата)`, которая возвращает название месяца на английском языке для указанной даты. Например, `MONTHNAME('2020-04-12')='April'`.
2. Если группировка осуществляется по вычисляемому столбцу (в данном случае «вычисляется» название месяца), то после `GROUP BY` можно указать как вычисляемое выражение, так и имя столбца, заданное с помощью `AS` .  
Важно отметить, что последний вариант (указать имя столбца)  нарушает стандарт по порядку выполнения запросов, но иногда может встречаться на реальных платформах.

In [59]:
%%sql
SELECT MONTHNAME(date_first) AS Месяц,
       COUNT(MONTH(date_first)) AS Количество
  FROM trip
 GROUP BY Месяц
 ORDER BY Количество DESC, Месяц

 * mysql://user:***@localhost:3306/stepik
7 rows affected.


[('February', 4),
 ('January', 4),
 ('June', 3),
 ('May', 3),
 ('April', 2),
 ('July', 2),
 ('March', 2)]

### Задание 9

Вывести сумму суточных (произведение количества дней командировки и размера суточных) для командировок,  
первый день которых пришелся на февраль или март 2020 года.  
Значение суточных для каждой командировки занесено в столбец `per_diem` .  
Вывести фамилию и инициалы сотрудника, город, первый день командировки и сумму суточных.  
Последний столбец назвать **Сумма** .  
Информацию отсортировать сначала в алфавитном порядке по фамилиям сотрудников, а затем по убыванию суммы суточных.

In [61]:
%%sql
SELECT name, city, date_first,
       per_diem * (DATEDIFF(date_last, date_first) + 1) AS Сумма
  FROM trip
 WHERE MONTH(date_first) IN (2,3)
   AND YEAR(date_first) = 2020
 ORDER BY name,
          Сумма DESC

 * mysql://user:***@localhost:3306/stepik
6 rows affected.


[('Абрамова К.А.', 'Москва', datetime.date(2020, 2, 23), Decimal('5600.00')),
 ('Баранов П.Е.', 'Москва', datetime.date(2020, 2, 14), Decimal('6300.00')),
 ('Колесов С.П.', 'Новосибирск', datetime.date(2020, 2, 27), Decimal('6750.00')),
 ('Колесов С.П.', 'Москва', datetime.date(2020, 2, 1), Decimal('4200.00')),
 ('Лебедев Т.К.', 'Москва', datetime.date(2020, 3, 3), Decimal('2800.00')),
 ('Семенов И.В.', 'Санкт-Петербург', datetime.date(2020, 3, 29), Decimal('5600.00'))]

### Задание 10

Только для этого задания изменена строка таблицы `trip`:

> | **4** | Ильиных Г.Р. | Владивосток | 450 | 2020-01-12 | **2020-03-02** |

(изначально: *2020-02-02*)

In [63]:
%%sql
UPDATE trip
   SET date_last = '2020-03-02'
 WHERE trip_id = 4

 * mysql://user:***@localhost:3306/stepik
1 rows affected.


[]

Вывести фамилию с инициалами и общую сумму суточных, полученных за все командировки для тех сотрудников,  
которые были в командировках больше чем 3 раза, в отсортированном по убыванию сумм суточных виде.  
Последний столбец назвать **Сумма**.

In [78]:
%%sql
SELECT name,
       SUM(per_diem * (DATEDIFF(date_last, date_first) + 1)) AS Сумма
  FROM trip
 GROUP BY name
HAVING COUNT(date_first) > 3
 ORDER BY Сумма DESC

 * mysql://user:***@localhost:3306/stepik
2 rows affected.


[('Абрамова К.А.', Decimal('29200.00')), ('Баранов П.Е.', Decimal('21300.00'))]